criterion-hooks.c

```diff
@@ -76,10 +80,26 @@ ReportHook(TEST_CRASH)(struct criterion_test_stats *stats) {
   }
 }
 
+static void process_failed_test(struct criterion_test_stats *test) {
+  // If there is a failed assertion, it most probably got already printed by assertion hook.
+  // Test has already been reported as failed.
+  for (struct criterion_assert_stats *assert = test->asserts; assert; assert = assert->next) {
+    if (!assert->passed) {
+      return;
+    }
+  }
+  print_failure_message(test->message, "Test failed");
```

Comment:
Honestly, I was wondering too, but I do not know. But I decided to include this branch, mostly because I could not find any documentation for Criterion hooks and lifecycle of tests, so I could not find any confirmation that the only possible way to fail a test is a failed assertion or expectation.

I decided to leave this branch simply because I do not reject a possibility that it's possible to have a test failing for some other reason than failed assertion. If you found it documented otherwise, or you know that it's not necessary, then maybe this part can be removed.

Reply:
Yeah, let's keep it until we find out.

Reply:
> Do you know how we can end up here? The test status is not `CR_STATUS_PASSED` and the test didn't time out either.

This line suggests it's possible to have a failed tests which has no failed assertions and is not timed out: https://github.com/Snaipe/Criterion/blob/a64b8600a9236a4cc8204353b19ce233e294c29b/src/core/stats.c#L237

```c
    if (test->failed_asserts > 0
            || test->timed_out
            || test->signal != test->test->data->signal
            || test->exit_code != test->test->data->exit_code)
        test->test_status = CR_STATUS_FAILED;
```

But still I am not sure what could lead to signals or exit codes mismatched between `test` and `test->test->data`.